share-extension/ShareViewController.m

```diff
@@ -169,11 +172,19 @@ -(void) getPostInfos {
                     if ([url isKindOfClass:[NSURL class]]){
                         postActivity.url = url;
                     } else if ([url isKindOfClass:[UIImage class]]){
-                        postActivity.fileData = UIImagePNGRepresentation((UIImage*)item);
+                        if ([itemProvider hasItemConformingToTypeIdentifier:(NSString *)kUTTypePNG]){
+                            postActivity.fileData = UIImagePNGRepresentation((UIImage*)item);
+                            postActivity.fileExtension = @"png";
+                        } else {
+                            postActivity.fileData = UIImageJPEGRepresentation((UIImage*)item, kJPEGCompressionLevel);
+                            postActivity.fileExtension = @"jpg";
+                        }
+                        
                     } else if ([url isKindOfClass:[NSData class]]){
                         postActivity.fileData = (NSData*)url;
+                        postActivity.fileExtension = [itemProvider hasItemConformingToTypeIdentifier:(NSString *)kUTTypePNG]?@"png": @"jpg";
```

Comment:
This line was found in the part where the itemProvider has type Image so it have to be an Image. Depending on the source of image the host app could send us a NSData or FileURL.